File: admin-panel/src/types/ingestion.ts
export type IngestionStatus =
  | 'UPLOADED'
  | 'MINERU_COMPLETE'
  | 'QUESTIONS_BUILT'
  | 'METADATA_COMPLETE'
  | 'SAVED'
  | 'FAILED'

export interface ProcessingReport {
  questions_detected: number
  questions_stored: number
  questions_skipped: number
  answers_mapped: number
  images_linked: number
  processing_time_s: number
  warnings: string[]
  errors: string[]
  per_question?: Array<{ number: number; status: string; warnings: string[] }>
}

export interface IngestionListItem {
  id: string
  filename: string
  status: IngestionStatus
  exam_name: string | null
  subject: string | null
  layout_type: string | null
  questions_saved: number
  created_at: string
}

export interface IngestionDetail extends IngestionListItem {
  error_message: string | null
  failed_at_stage: string | null
  questions_found: number
  detected_layout: string | null
  detected_subjects: string[]
  processing_time_s: number | null
  processing_report: ProcessingReport | null
  mineru_output_dir: string | null
}

File: app/schemas/ingestion.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, field_validator

from app.services.question_builder.registry import VALID_LAYOUT_TYPES

# Allowed subject values — extensible list
VALID_SUBJECTS = [
    "physics", "chemistry", "mathematics", "biology",
    "english", "hindi", "history", "geography",
]


class IngestionCreate(BaseModel):
    filename: str


class IngestionStatusResponse(BaseModel):
    id: uuid.UUID
    filename: str
    status: str
    # Upload-form fields
    exam_name: str | None = None
    subject: str | None = None
    layout_type: str | None = None
    # State
    error_message: str | None = None
    failed_at_stage: str | None = None
    questions_found: int
    questions_saved: int
    detected_layout: str | None = None
    detected_subjects: list[str] = []
    processing_time_s: float | None = None
    processing_report: dict | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class IngestionListItem(BaseModel):
    id: uuid.UUID
    filename: str
    status: str
    exam_name: str | None = None
    subject: str | None = None
    layout_type: str | None = None
    questions_saved: int
    created_at: datetime

    model_config = {"from_attributes": True}
